File: server/src/algo2_py/box_motion.py
from box import get_size
from enum import IntEnum
import random


class Rotation(IntEnum):
    WHL = 0
    LHW = 1
    HLW = 2
    LWH = 3
    WLH = 4
    HWL = 5


def rotate_each_box(boxes):
    numOfRotations = len(Rotation)
    rotationsList = list(Rotation)
    return [
        {**box, "rotation": random.choice(rotationsList)}
        for box in boxes
    ]


def rotate_subset(boxes):
    """
    For each subset of items sharing all dimensions,
    randomly pick one of the orientations that with equal probability.
    """

    # Group boxes by size
    sameSizeDict = {}
    for box in boxes:
        boxSizeString = str(get_size(box))
        if boxSizeString not in sameSizeDict:
            sameSizeDict[boxSizeString] = []
        sameSizeDict[boxSizeString].append(box)

    # Generate random rotation for each size
    numOfRotations = len(Rotation)
    rotationsList = list(Rotation)
    rotations = {
        size: random.choice(rotationsList)
        for size in sameSizeDict.keys()
    }

    # Assign rotation to boxes
    return [
        {**box, "rotation": rotations[str(get_size(box))]}
        for box in boxes
    ]


def rotation(boxes):
    # Rotate boxes individually or by subset with equal probability
    if len(boxes):
        return (
            rotate_each_box(boxes)
            if random.random() < 0.5
            else rotate_subset(boxes)
        )
    else:
        return []


def volume_perturb(b1, b2):
    if 0.7 <= b1["volume"] / b2["volume"] <= 1.3 and random.random() > 0.5:
        return True
    return False


def perturbation(boxes):
    if len(boxes) <= 1:
        return boxes
    for i in range(len(boxes) - 1):
        if volume_perturb(boxes[i], boxes[i + 1]):
            boxes[i], boxes[i + 1] = boxes[i + 1], boxes[i]

    return boxes

File: server/src/algorithm/algo_js/boxMotion.js
import { getBoxRotatedSize, getValidRotations } from "./box.js";
import { Rotation } from "./utils.js";

const rotateEachBox = (boxes, container) => {
    return boxes.map((box) => {
        const validRotations = getValidRotations(box, container);
        const numOfRotations = Object.values(validRotations).length;
        const rotation =
            validRotations[Math.floor(Math.random() * numOfRotations)];
        return { ...box, rotation: rotation };
    });
};

const rotateSubset = (boxes, container) => {
    /*
    For each subset of items sharing all dimensions,
    randomly pick one of the orientations that with equal probability.
    */

    // Group boxes by size
    const sameSizeDict = boxes.reduce((dict, box) => {
        const boxSizeString = JSON.stringify(getBoxRotatedSize(box));
        if (!dict[boxSizeString]) {
            dict[boxSizeString] = [];
        }
        dict[boxSizeString].push(box);
        return dict;
    }, {});

    // Generate random rotation for each size
    const rotations = Object.keys(sameSizeDict).reduce((acc, size) => {
        if (!acc[size]) {
            const validRotations = getValidRotations({ size }, container);
            const numOfRotations = Object.values(validRotations).length;
            const rotationIndex = Math.floor(Math.random() * numOfRotations);
            const rotation = validRotations[rotationIndex];
            acc[size] = rotation;
        }
        return acc;
    }, {});

    // Assign rotation to boxes
    return boxes.map((box) => {
        const boxSizeString = JSON.stringify(getBoxRotatedSize(box));
        const rotation = rotations[boxSizeString];
        return { ...box, rotation: rotation };
    });
};

export const rotation = (boxes, container) => {
    // Rotate boxes individually or by subset with equal probability
    if (boxes.length) {
        return Math.random() < 0.5
            ? rotateEachBox(boxes, container)
            : rotateSubset(boxes, container);
    } else {
        return [];
    }
};

const volumePerturb = (b1, b2) => {
    if (
        0.7 <= b1.volume / b2.volume &&
        b1.volume / b2.volume <= 1.3 &&
        Math.random() > 0.5
    ) {
        return true;
    }
    return false;
};

export const perturbation = (boxes) => {
    if (boxes.length <= 1) {
        return boxes;
    }
    for (let i = 0; i < boxes.length - 1; i++) {
        if (volumePerturb(boxes[i], boxes[i + 1])) {
            [boxes[i], boxes[i + 1]] = [boxes[i + 1], boxes[i]];
        }
    }

    return boxes;
};
